src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGeneratorHelper.cs

```diff
@@ -0,0 +1,559 @@
+        private bool InitializeContextClass(Type type, string className, StringBuilder source)
+        {
+            source.Append($@"
+
+namespace {_generationNamespace}
+{{
+    public partial class JsonContext : JsonSerializerContext
+    {{
+        private {className}TypeInfo _{className};
+        public JsonTypeInfo<{type.Name}> {className} 
+        {{
+            get
+            {{
+                if (_{className} == null)
+                {{
+                    _{className} = new {className}TypeInfo(this);
+                }}
+
+                return _{className}.TypeInfo;
+            }}
+        }}
+        ");
+
+            return true;
+        }
+
+        private bool InitializeTypeClass(string className, StringBuilder source) {
+            source.Append($@"
+        private class {className}TypeInfo 
+        {{
+        ");
+
+            return true;
+        }
+
+        private bool TypeInfoGetterSetter(Type type, StringBuilder source)
+        {
+            source.Append($@"
+            public JsonTypeInfo<{type.Name}> TypeInfo {{ get; private set; }}
+            ");
+
+            return true;
+        }
+
+        private bool InitializeTypeInfoProperties(PropertyInfo[] properties, StringBuilder source)
+        {
+            Type genericType;
+            string typeName;
+            string propertyName;
+
+            foreach (PropertyInfo property in properties)
+            {
+                // todo: Verify if the type has already failed in sourcegeneration.
+
+                // Find type and property name to use for property definition.
+                typeName = property.PropertyType.Name;
+                propertyName = property.Name;
+                if (property.PropertyType is TypeWrapper type)
+                {
+                    // Check if IEnumerable.
+                    if (type.IsIEnumerable)
+                    {
+                        genericType = GetTypesToGenerate(type).First();
+                        if (type.IsIList)
+                        {
+                            typeName = $"List<{genericType.Name}>";
+                        }
+                        else
+                        {
+                            // todo: Add support for rest of the IEnumerables.
+                            return false;
+                        }
+                    }
+                }
+
+                source.Append($@"
+            private JsonPropertyInfo<{typeName}> _property_{propertyName};
+                ");
+            }
+
+            return true;
+        }
+
+        private bool GenerateTypeInfoConstructor(Type root, string className, PropertyInfo[] properties, FieldInfo[] fields, StringBuilder source)
+        {
+            source.Append($@"
+            public {className}TypeInfo(JsonContext context)
+            {{
+                var typeInfo = new JsonObjectInfo<{root.Name}>(CreateObjectFunc, SerializeFunc, DeserializeFunc, context.GetOptions());
```

Comment:
> root.Name

this won't work if two types have the same name.

Reply:
Changed to assembly+fullname of the type :)